src/Builder/FrontendDialectHelper.cpp

```diff
@@ -11,232 +11,203 @@
+using DenseElementsAttrBuilder =
+    llvm::function_ref<mlir::DenseElementsAttr(llvm::ArrayRef<T>)>;
 
-// Helper method for constructing an array attribute from a model input.
+// Returns DenseElementsAttr with tp's data.
 template <typename T>
-std::vector<T> CreateArrayAttribute(onnx::TensorProto initializer) {
-  // Return an emtpy array if the raw data is empty.
-  if (initializer.dims().size() == 1 && initializer.dims().data()[0] == 0)
-    return std::vector<T>();
-
-  size_t size;
-  if (initializer.raw_data().size()) {
-    // Copy & take care of endianness.
-    std::vector<char> byteInitializer;
-    std::copy(initializer.raw_data().begin(), initializer.raw_data().end(),
-        back_inserter(byteInitializer));
-    size = initializer.raw_data().size() / sizeof(T);
-    T *arrayPtr = reinterpret_cast<T *>(&byteInitializer[0]);
-    auto array = std::vector<T>(arrayPtr, arrayPtr + size);
+mlir::DenseElementsAttr createDenseElmAttr(const std::string &externalDataDir,
+    const onnx::TensorProto &tp, DenseElementsAttrBuilder<T> denseBuilder) {
+  std::unique_ptr<llvm::MemoryBuffer> externalData =
+      (tp.has_data_location() &&
+          tp.data_location() == onnx::TensorProto::EXTERNAL)
+          ? readExternalData(externalDataDir, tp)
+          : nullptr;
+  if (externalData || tp.has_raw_data()) {
+    llvm::StringRef buffer = externalData ? externalData->getBuffer()
+                                          : llvm::StringRef(tp.raw_data());
+    size_t size = buffer.size() / sizeof(T);
+    llvm::ArrayRef<T> arrayRef(
+        reinterpret_cast<T const *>(buffer.data()), size);
     // Perform byte swap if system endianness is BE.
     // ONNX tensor content raw data is always in LE.
-    if (llvm::support::endian::system_endianness() !=
-        llvm::support::endianness::little)
-      for (size_t i = 0; i < array.size(); i++)
-        llvm::sys::swapByteOrder<T>(array[i]);
-
-    return array;
+    if (sizeof(T) > 1 && llvm::support::endian::system_endianness() !=
+                             llvm::support::endianness::little) {
+      llvm::SmallVector<T> vector;
+      vector.reserve(size);
+      for (T x : arrayRef) {
+        vector.push_back(llvm::sys::getSwappedBytes(x));
+      }
+      return denseBuilder(llvm::makeArrayRef(vector));
+    } else {
+      // No need to take care of endianness.
+      return denseBuilder(arrayRef);
+    }
+  } else {
+    // Not raw, no need to take care of endianness.
+    const auto &data = TransformValueToONNXData<T>::data(tp);
+    // Access data directly via ArrayRef if same size as T,
+    // or copy into correctly typed SmallVector otherwise
+    // because DenseElementsAttr needs argument type of the correct bitwidth.
+    typedef typename std::conditional<sizeof(T) == sizeof(data[0]),
+        llvm::ArrayRef<T>, llvm::SmallVector<T>>::type ArrayRefOrSmallVector;
+    ArrayRefOrSmallVector array(data.begin(), data.end());
+    return denseBuilder(llvm::makeArrayRef(array));
```

Comment:
avoiding a memory copy here in the cases where T has the same size as data's type's value_type, namely for float, double, i32, i64

the cases that have to do a memory copy via SmallVector are bool, i8, u8, i16, u16 from tp.int32_data, and u32 from tp.uint64_data